src/network-manager.js

```diff
@@ -0,0 +1,281 @@
+const Platform = require('./platform');
+
+function stripQuotes(str) {
+  return str.replace(/^'+/, '').replace(/'+$/, '');
+}
+
+/**
+ * Get the WAN mode and options.
+ *
+ * @returns {Object} {mode: 'static|dhcp|pppoe|...', options: {...}}
+ */
+function getWanMode() {
+  let mode = null;
+  const options = {};
+
+  switch (Platform.getOS()) {
+    case 'linux-openwrt': {
+      const proc = child_process.spawnSync(
+        'uci',
+        ['show', 'network.wan'],
+        {encoding: 'utf8'}
+      );
+      if (proc.status !== 0) {
+        return {mode, options};
+      }
+
+      for (const line of proc.stdout.split('\n')) {
+        if (line.indexOf('=') < 0) {
+          continue;
+        }
+
+        const [key, value] = line.split('=', 2).map((s) => stripQuotes(s));
+
+        // discard network.wan=interface
+        if (!key.startsWith('network.wan.')) {
+          continue;
+        }
+
+        const opt = key.split('network.wan.')[1];
+        if (opt === 'proto') {
+          mode = value;
+        } else {
+          options[opt] = value;
+        }
+      }
+
+      return {mode, options};
+    }
+    default:
+      return {mode, options};
+  }
+}
+
+/**
+ * Set the WAN mode and options.
+ *
+ * @param {string} mode - static, dhcp, pppoe, ...
+ * @param {Object?} options - options specific to WAN mode
+ * @returns {boolean} Boolean indicating success.
+ */
+function setWanMode(mode, options = {}) {
+  const valid = ['static', 'dhcp', 'pppoe'];
+  if (!valid.includes(mode)) {
+    return false;
+  }
+
+  switch (Platform.getOS()) {
+    case 'linux-openwrt': {
+      let proc = child_process.spawnSync(
+        'uci',
+        ['set', `network.wan.proto=${mode}`]
+      );
+      if (proc.status !== 0) {
+        return false;
+      }
+
+      for (const [key, value] of Object.entries(options)) {
+        proc = child_process.spawnSync(
+          'uci',
+          ['set', `network.wan.${key}=${value}`]
+        );
+        if (proc.status !== 0) {
+          return false;
+        }
+      }
+
+      proc = child_process.spawnSync('uci', ['commit', 'network']);
+      if (proc.status !== 0) {
+        return false;
+      }
+
+      proc = child_process.spawnSync('/etc/init.d/network', ['reload']);
+      return proc.status === 0;
+    }
+    default:
+      return false;
+  }
+}
+
+function getWirelessMode() {
+  let enabled = false;
+  let mode = null;
+  const options = {};
+
+  switch (Platform.getOS()) {
+    case 'linux-openwrt': {
+      let proc = child_process.spawnSync(
+        'uci',
+        ['show', 'wireless'],
+        {encoding: 'utf8'}
+      );
+      if (proc.status !== 0) {
+        return {enabled, mode, options};
+      }
+
+      let device = null;
+      let iface = null;
+      for (const line of proc.stdout.split('\n')) {
+        if (line.indexOf('=') < 0) {
+          continue;
+        }
+
+        const [key, value] = line.split('=', 2).map((s) => stripQuotes(s));
+
```

Comment:
It feels like everything from spawnSync to here could be encapsulated into a common function?
Perhaps something that calls uci show and splits the results into lines?

Reply:
Agreed, this could be cleaned up a bit.